src/vmemcache.c

```diff
@@ -98,6 +439,46 @@ vmemcache_get(VMEMcache *cache, const char *key, size_t ksize, void *vbuf,
 int
 vmemcache_evict(VMEMcache *cache, const char *key, size_t ksize)
 {
+	struct cache_entry *entry;
+	int evicted = 0;
+
+	if (key == NULL) {
+		entry = cache->repl.ops->repl_p_evict(cache->repl.head, NULL);
+		if (entry == NULL) {
+			ERR("no element to evict");
+			errno = EINVAL;
+			return -1;
+		}
+
+		entry->value.p_entry = NULL;
+		evicted = 1;
+
+		key = entry->key.key;
+		ksize = entry->key.ksize;
+	}
+
+	if (cache->on_evict != NULL)
+		(*cache->on_evict)(cache, key, ksize, cache->arg_evict);
+
+	int ret = vmemcache_find_key_remove(cache, key, ksize,
+					&entry, 1 /* remove it */);
+	if (ret < 0 || entry == NULL) {
+		ERR("cannot find an element with the given key");
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (!evicted) {
+		cache->repl.ops->repl_p_evict(cache->repl.head,
+						entry->value.p_entry);
+		entry->value.p_entry = NULL;
+	}
+
+	if (__sync_fetch_and_sub(&entry->value.refcount, 1) == 1 &&
```

Comment:
wrap this code into separate function (entry_refcount_inc/refcount_dec?)

Reply:
Done